src/video.cpp

```diff
@@ -5,861 +5,908 @@
+	_window_manager.KeyEvent(key_code, mod_mask, symbol);
 }
 
 /**
- * Mark the stated area of the screen as being out of date.
- * @param rect %Rectangle which is out of date.
- * @todo Keep an administration of the rectangle(s??) to update, and update just that part.
+ * Called by OpenGL when text is entered.
+ * @param window Window in which the event occurred.
+ * @param utf32 The UTF-32 encoded character.
  */
-void VideoSystem::MarkDisplayDirty([[maybe_unused]] const Rectangle32 &rect)
+void VideoSystem::TextCallback(GLFWwindow *window, uint32 utf32)
 {
-	this->dirty = true;
+	assert(window == _video.window);
+	char buffer[] = {0, 0, 0, 0, 0};
+	EncodeUtf8Char(utf32, buffer);
+	_window_manager.KeyEvent(WMKC_SYMBOL, WMKM_NONE, buffer);
 }
 
-/** Mark the display as being up-to-date. */
-void VideoSystem::MarkDisplayClean()
+/**
+ * Called by OpenGL when the mouse was moved.
+ * @param window Window in which the event occurred.
+ * @param x New mouse position X coordinate.
+ * @param y New mouse position Y coordinate.
+ */
+void VideoSystem::MouseMoveCallback(GLFWwindow *window, double x, double y)
 {
-	this->dirty = false;
+	assert(window == _video.window);
+	_video.mouse_x = std::max(0.0, std::min<double>(x, _video.width));
+	_video.mouse_y = std::max(0.0, std::min<double>(y, _video.height));
+	_window_manager.MouseMoveEvent();
 }
 
 /**
- * Set the clipped area.
- * @param cr New clipped blitting area.
+ * Called by OpenGL when the mouse wheel was moved.
+ * @param window Window in which the event occurred.
+ * @param xdelta Amount by which the wheel was moved in the horizontal direction..
+ * @param ydelta Amount by which the wheel was moved in the vertical direction.
  */
-void VideoSystem::SetClippedRectangle(const ClippedRectangle &cr)
+void VideoSystem::ScrollCallback(GLFWwindow *window, [[maybe_unused]] double xdelta, double ydelta)
 {
-	this->blit_rect = cr;
+	assert(window == _video.window);
+	if (abs(ydelta) < 0.01) return;
+	_window_manager.MouseWheelEvent((ydelta > 0) ? 1 : -1);
 }
 
 /**
- * Get the current clipped blitting area.
- * @return Current clipped area.
+ * Called by OpenGL when a mouse click occurs.
+ * @param window Window in which the event occurred.
+ * @param button Mouse button constant.
+ * @param action Whether the mouse button was pressed or released.
+ * @param mods Bitset of modifier keys.
  */
-ClippedRectangle VideoSystem::GetClippedRectangle()
-{
-	this->blit_rect.ValidateAddress();
-	return this->blit_rect;
+void VideoSystem::MouseClickCallback(GLFWwindow *window, int button, int action, [[maybe_unused]] int mods) {
+	assert(window == _video.window);
+
+	switch (button) {
+		case GLFW_MOUSE_BUTTON_RIGHT:
+			_window_manager.MouseButtonEvent(MB_RIGHT, action == GLFW_PRESS);
+			break;
+		case GLFW_MOUSE_BUTTON_LEFT:
+			_window_manager.MouseButtonEvent(MB_LEFT, action == GLFW_PRESS);
+			break;
+		case GLFW_MOUSE_BUTTON_MIDDLE:
+			_window_manager.MouseButtonEvent(MB_MIDDLE, action == GLFW_PRESS);
```

Comment:
I have seen people use `GLFW_RELEASE` here, allowing you to either press at the wrong spot and move towards it before release, or have a change of mind after pressing, and move away before release.

Mostly useful for non-mouse systems though, so it's fine to keep this now imho.

Reply:
This would have to be implemented in the WindowManager. This is actually not trivial because it breaks movement initiation. Maybe something for a future branch.